pypiserver/_app.py

```diff
@@ -289,14 +298,55 @@ def simple(project):
     if project != normalized:
         return redirect(f"/simple/{normalized}/", 301)
 
+    # pip install always uses 2 requests, first to get the list of versions
+    # and then to download the package.
+    # If we only search for the package on the local filesystem, we will
+    # might not be able to return the list containing the version.
+    # Hence, if pull_through is enabled, we will fetch the version list
+    # from pypi.org and then return the list of versions.
+
+    # If the packages version is not found, in pypi.org, we will return a 404
+
+    # Serving the actual package is handled by the /packages/ route
+    # And in this route, we will first attempt to fetch the package from the
+    # local filesystem, if not found, we will fetch it from pypi.org and save
+    # it to the local filesystem and then serve it.
+    if config.pull_through:
+        # Getting a list of packages and hashes
+        rgx = re.compile(
+            r"href=\"(?P<url>.*)#sha256=(?P<meta>.*\") ?>(?P<pkgversion>.*)</a><br />"
+        )
+        res = urllib.request.urlopen(
+            f"{config.fallback_url.rstrip('/')}/{project}/"
+        )
+        content = res.read().decode("utf-8").split("\n")
+        new_content = []
+        for line in content:
+            m = rgx.search(line)
+            if m:
+                gdict = m.groupdict()
+                vrsn = gdict["pkgversion"]
+                meta = gdict["meta"]
+                val = f'<a href="http://{config.host}:{config.port}/packages/{vrsn}#{meta}>{vrsn}</a><br />'
+            else:
+                val = line
+            new_content.append(val)
+        body = "\n".join(new_content)
+        response = HTTPResponse(body=body, status=200)
+        return response
```

Comment:
[curious] do I correctly understand that in case we've enabled `pull_through`, the simple request returns only the masked URLs from the fallback URL? And then we implicitly rely on retrieving them from the fallback or local. Please correct me if I misunderstand. 

I wonder wouldn't it actually be more valuable to return a union of the URLs: resolved locally U masked from the pull-through setting. What do you think? 

In my opinion we'll then return a more complete set. Iirc, we can't guarantee that versions uploaded locally (suppose someone's pushing forked package versions but falls back to a set remote) and those stored on the fallback are the same? Sort of a `FULL OUTER JOIN` situation :)   

But let me know if I'm mistaken please :v:

Reply:
Basically, you are correct. 
If you ask for a package without a version than the correct thing would be to present the client with a union of packages found locally and upstream.
But if you point the client for fetching a package upstream you will not be be able to do the pull-through trick.

(Also, usually, clients do ask for packages with versions)

Example dependancy graph:

A->B->C

You do pip install A.

pypiserver receives a request for A.

A isn't found locally.

Pypiserver fetches the metadata and presenta it pip .

pip determines that version 12.3.4 is the latest version and that there is wheel.
It sends back a request to pypiserver requesting A-12.3.4.whl.

pypiserver does not find it locally so it fetches the file from upstream and saves it locally, then it serves it to pip.

Pip downloads the wheel, installs it, the sees that it needs to ask for B in version 9.8.7. The cycle goes on again ...,

Reply:
Right, not that sounds all good. What I mean is more regarding this: 

> If you ask for a package without a version than the correct thing would be to present the client with a union of packages found locally and upstream.

> But if you point the client for fetching a package upstream you will not be be able to do the pull-through trick.

The part I am concerned about is not the masking of the urls as I understand that in the pull-through case, we'll eventually fetch them too. What I mean is that here it seems we return (line 336) _only_ the remote URLs masked, without the local packages at all. 

So suppose package `A` has: 
1. locally: `A:0.0.1`, `A:0.0.2`, and ``A:0.0.2-rc1`
2. remotely: `A:0.0.1`, `A:0.0.2`, and `A:0.0.3`

Then in my opinion the output of the `/simple/A` should be as follows: 
- `A:0.0.1` 
- `A:0.0.2`
- `A:0.0.2-rc1`
    > this is only on local
- `A:0.0.3` 
    > this appears as if on local, but will be pulled-through from the remote

Afaiu, current code with `pull-through` enabled, will return only the `A:0.0.1`, `A:0.0.2`, and `A:0.0.3`. Is that correct?
Does it make sense? 🤔 

Let me know :v: